docs: tutorials introduction

jean-michelet opened the pull request:
See https://github.com/fastify/fastify/issues/6149

This is currently a draft intended to gather feedback.

I think it's best to introduce all the concepts related to routing, hooks, error handling, decorators and so on before plugins registration and encapsulation. Because these are concepts learners need to understand to fully grasp the power of encapsulation.

Once the first four sections are in place, we can probably consider linking it to the `fastify/website`. The TypeScript section can come later.

One thing I’m unsure about: the `docs` folder [is included in npm packages](https://www.npmjs.com/package/fastify?activeTab=code). I think Matteo does this so users always have a copy of the documentation, even if the original repository disappears. But I’m not sure if that should apply to this tutorial specifically. 

Maybe it should live somewhere else?


## 💰 Bounty

This issue has a bounty of $800.
_Read more about [the bounty program](./EXPENSE_POLICY.md)_

jean-michelet commented:
Hi!

I was very busy at work, but now I can continue to work on this.
Just need more feedback plz.

mcollina commented:
Keep them in this PR. Youl could also do PR using this as base, to reduce the volume of reviews.

gurgunday reviewed (commented):
I think this would help a lot of people

lgtm so far!

jean-michelet commented:
We can create a PR by big theme, here is the introduction to the tutorial and basics.

jean-michelet commented on docs/Tutorial/3-create-server.md:
````diff
@@ -0,0 +1,115 @@
+
+You should see a message indicating your Fastify server is running on port 3000.
+
+### Fastify Internals
+
+The instance returned by the factory function provides access to public 
+methods like `listen()`, which we used above. 
+You can explore the source code of the factory function 
+[here](https://github.com/fastify/fastify/blob/ad97fbb5630488b2c830ff7bc27f495b21b87243/fastify.js#L102)
+. The `listen` method is set
+[here](https://github.com/fastify/fastify/blob/ad97fbb5630488b2c830ff7bc27f495b21b87243/fastify.js#L338).
+
+Internally, Fastify manages its state using symbol-defined properties to avoid 
+collisions and distinguish them from public properties. Symbol identifiers 
+typically have a `k` prefix. For instance, the internal state is `kState`:
+
+```javascript
+const fastify = {
+  [kState]: {
+    listening: false,
+    closing: false,
+    started: false,
+    ready: false,
+    booting: false,
+    aborted: false,
+    readyResolver: null
+  },
+  // ...additional internals
+}
+```
+
+Explore all internal symbols defined by Fastify 
+[here](https://github.com/fastify/fastify/blob/main/lib/symbols.js).
````
In my experience, developers often overestimate the complexity of the tools they use and think they are not qualified to contribute. 

By gradually introducing internal mechanisms, without going too deep into the details of implementation, I hope to encourage readers to think about the code that runs in the background and understand that contributing to Fastify is both accessible and useful.

jean-michelet replied:
This would be facultative sections at the end of each chapter.

Wdyt?

Uzlopak committed:
fix: handle non FastifyErrors in custom handler properly, set type of error-parameter for setErrorHandler and errorHandler to unknown, but configurable via generic TError (#6308)

* types: loosen setErrorHandler and errorHandler types

* set TError to unknown
add generic to errorHandler

* fix

* remove obsolte type tst

* add unit test, fix potential issue

---------

Co-authored-by: Jean <110341611+jean-michelet@users.noreply.github.com>

build(deps-dev): remove @fastify/pre-commit (#6319)

`@fastify/pre-commit` isn't configured or used in this repo as there is no `pre-commit` object in package.json. It won't run regardless as scripts are disabled.

Signed-off-by: Frazer Smith <frazer.dev@icloud.com>
Co-authored-by: Aras Abbasi <aras.abbasi@googlemail.com>

ci: improve citgm workflows (#6334)

* ci: improve citgm workflows

* fix remark

docs: explain stream error handling (#5746)

* docs: explain stream error handling

* Update docs/Reference/Server.md

Co-authored-by: Matteo Collina <matteo.collina@gmail.com>
Signed-off-by: James Sumners <321201+jsumners@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Frazer Smith <frazer.dev@icloud.com>
Signed-off-by: Denys Otrishko <shishugi@gmail.com>

* docs: explain stream error handling

* fix test

---------

Signed-off-by: James Sumners <321201+jsumners@users.noreply.github.com>
Signed-off-by: Denys Otrishko <shishugi@gmail.com>
Signed-off-by: Aras Abbasi <aras.abbasi@googlemail.com>
Co-authored-by: James Sumners <321201+jsumners@users.noreply.github.com>
Co-authored-by: Matteo Collina <matteo.collina@gmail.com>
Co-authored-by: Frazer Smith <frazer.dev@icloud.com>
Co-authored-by: Aras Abbasi <aras.abbasi@googlemail.com>

fix: error throwing in reply (#6299)

* fix: instantiate readable stream error

* Create reply-web-stream-locked.test.js

Signed-off-by: Juan L. <juanf03@gmail.com>
Signed-off-by: Juan Letamendia <juanf03@gmail.com>

* fix test

---------

Signed-off-by: Juan L. <juanf03@gmail.com>
Signed-off-by: Juan Letamendia <juanf03@gmail.com>
Co-authored-by: Aras Abbasi <aras.abbasi@googlemail.com>

docs: create 2 frst chapters of the tutorial

docs: nit

docs: create a fastify server

docs: create basic  routes

docs: improve chapter 4

refactor: improve readability and quote examples

docs: decoration chapter public part

docs: decoration chapter public part

fix: typo

fix: typo

fix: markdown lint

docs: document internals

docs: validatio nand serialization

docs: move section mapping reminder with routes declarations

docs: add hooks chapter

docs: remove internals involved sections

docs: add error handling chapter

fix: copilot suggestions

Uzlopak committed:
fix: handle non FastifyErrors in custom handler properly, set type of error-parameter for setErrorHandler and errorHandler to unknown, but configurable via generic TError (#6308)

* types: loosen setErrorHandler and errorHandler types

* set TError to unknown
add generic to errorHandler

* fix

* remove obsolte type tst

* add unit test, fix potential issue

---------

Co-authored-by: Jean <110341611+jean-michelet@users.noreply.github.com>

build(deps-dev): remove @fastify/pre-commit (#6319)

`@fastify/pre-commit` isn't configured or used in this repo as there is no `pre-commit` object in package.json. It won't run regardless as scripts are disabled.

Signed-off-by: Frazer Smith <frazer.dev@icloud.com>
Co-authored-by: Aras Abbasi <aras.abbasi@googlemail.com>

ci: improve citgm workflows (#6334)

* ci: improve citgm workflows

* fix remark

docs: explain stream error handling (#5746)

* docs: explain stream error handling

* Update docs/Reference/Server.md

Co-authored-by: Matteo Collina <matteo.collina@gmail.com>
Signed-off-by: James Sumners <321201+jsumners@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Frazer Smith <frazer.dev@icloud.com>
Signed-off-by: Denys Otrishko <shishugi@gmail.com>

* docs: explain stream error handling

* fix test

---------

Signed-off-by: James Sumners <321201+jsumners@users.noreply.github.com>
Signed-off-by: Denys Otrishko <shishugi@gmail.com>
Signed-off-by: Aras Abbasi <aras.abbasi@googlemail.com>
Co-authored-by: James Sumners <321201+jsumners@users.noreply.github.com>
Co-authored-by: Matteo Collina <matteo.collina@gmail.com>
Co-authored-by: Frazer Smith <frazer.dev@icloud.com>
Co-authored-by: Aras Abbasi <aras.abbasi@googlemail.com>

fix: error throwing in reply (#6299)

* fix: instantiate readable stream error

* Create reply-web-stream-locked.test.js

Signed-off-by: Juan L. <juanf03@gmail.com>
Signed-off-by: Juan Letamendia <juanf03@gmail.com>

* fix test

---------

Signed-off-by: Juan L. <juanf03@gmail.com>
Signed-off-by: Juan Letamendia <juanf03@gmail.com>
Co-authored-by: Aras Abbasi <aras.abbasi@googlemail.com>

docs: create 2 frst chapters of the tutorial

docs: nit

docs: create a fastify server

docs: create basic  routes

docs: improve chapter 4

refactor: improve readability and quote examples

docs: decoration chapter public part

docs: decoration chapter public part

fix: typo

fix: typo

fix: markdown lint

docs: document internals

docs: validatio nand serialization

docs: move section mapping reminder with routes declarations

docs: add hooks chapter

docs: remove internals involved sections

docs: add error handling chapter

fix: copilot suggestions

jean-michelet commented:
Oh no, I closed the wrong PR.

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet committed:
fix: revert runtime file changes

jean-michelet committed:
fix: copilot feedback

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet commented:
Can't this do all the feedbacks at one lol

Uzlopak commented:
I really dont know. But why dont you just commit it directly with the button it provides?

jean-michelet committed:
docs: minor improvements

jean-michelet committed:
fix: nit

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

Eomm commented on docs/Tutorial/4-defining-routes.md:
```diff
@@ -0,0 +1,234 @@
+
+Under the hood, Fastify uses [find-my-way](https://github.com/delvedor/find-my-way),
+a very fast, framework-independent HTTP router that relies on a highly efficient
+[**Radix Tree**](https://en.wikipedia.org/wiki/Radix_tree)
```
Why does the reader care?

We should specify that the documentation of that module may be useful to understand how to setup complex path parameter or so

Eomm commented on docs/Tutorial/4-defining-routes.md:
```diff
@@ -0,0 +1,234 @@
+Fastify provides multiple ways to define routes:
+
+* **Full Declaration**: Explicitly declare routes using `fastify.route`,
+  see [documentation](https://fastify.dev/docs/latest/Reference/Routes/#full-declaration).
```
if we use relative path, the website should work out of the box

Eomm reviewed (commented):
Not read it all or run the code - but the backbone is there 👍🏼 🥇

jean-michelet commented:
> Oh, forgot to mention that we may think to move this content to the website directly instead?
> Just one option if we want more cool features

What feature do you think would be useful?

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet committed:
docs: typo

Co-authored-by: Manuel Spigolon <behemoth89@gmail.com>
Signed-off-by: Jean <110341611+jean-michelet@users.noreply.github.com>

jean-michelet committed:
docs: fundation link

Co-authored-by: Manuel Spigolon <behemoth89@gmail.com>
Signed-off-by: Jean <110341611+jean-michelet@users.noreply.github.com>

jean-michelet committed:
docs: typo

Co-authored-by: Manuel Spigolon <behemoth89@gmail.com>
Signed-off-by: Jean <110341611+jean-michelet@users.noreply.github.com>

jean-michelet committed:
docs: default port 0.0.0.0

Co-authored-by: Manuel Spigolon <behemoth89@gmail.com>
Signed-off-by: Jean <110341611+jean-michelet@users.noreply.github.com>

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet committed:
fic: eomm feedback

jean-michelet committed:
fix: markdown line limit

jean-michelet committed:
docs: write plugins chapter

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet committed:
docs: awaiting a plugin should be part of the boostraping section

jean-michelet committed:
docs: add chapter on testing

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet committed:
fix: markdown lint

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

bbberns commented:
## Sovereign OS: High-Resolution Optimization
Documentation improvement proposed for project stability. Payout to LTC: ltc1qx2g4z3uk2gzux77k8gkyd0aduzfx30rane4gds

Eomm committed:
fix: tutorials feedback (#1)

* chore: rename file number

* fix: chapter 1&2 feedback

* fix: chapter 3 feedback

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

robinshen36 commented:
Hi! I'd like to work on this task. I have the relevant skills and can deliver quality work. Please assign this to me!

aglichandrap commented:
## Bounty Claim

Hi! I'd like to work on this documentation bounty.

**My approach:**
1. Read the existing codebase and documentation
2. Identify gaps and areas for improvement
3. Write comprehensive, clear documentation
4. Include code examples where appropriate

**Timeline:** 24-48 hours

Looking forward to contributing!

Fdawgs commented:
@jean-michelet Probs best to remove mention of the concept that shares the same name as a chocolate bar, as it attracts bots that specifically hunt for it:
<img width="325" height="143" alt="image" src="https://github.com/user-attachments/assets/c6205a63-b42e-4cdc-97a6-1d78b232a361" />

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet commented:
@Fdawgs 
Initially, I was thinking to give the Bounty to one of us or a legit Node expert.
I am very busy at the moment and can't finish this, someone want to give it a try @fastify/core?

If no, I should be able to continue during the summer.

ai-hustle-bro commented:
I checked the current `linkChecker` failure. It looks like all 9 failures are internal docs links in the new Tutorial files using `/docs/latest/Reference/...`, which the workflow resolves as local files without the `.md` filename.

Concrete replacements that should clear the current failures:

```text
docs/Tutorial/03-create-server.md
- /docs/latest/Reference/Server/#factory
+ ../Reference/Server.md#factory

docs/Tutorial/04-defining-routes.md
- /docs/latest/Reference/Routes/#full-declaration
+ ../Reference/Routes.md#full-declaration
- /docs/latest/Reference/Routes/#async-await
+ ../Reference/Routes.md#async-await
- /docs/latest/Reference/Errors/#fst_err_ctp_invalid_media_type
+ ../Reference/Errors.md#fst_err_ctp_invalid_media_type

docs/Tutorial/05-decoration.md
- /docs/latest/Reference/Request/
+ ../Reference/Request.md
- /docs/latest/Reference/Reply/
+ ../Reference/Reply.md

docs/Tutorial/06-validation-and-serialization.md
- /docs/latest/Reference/Validation-and-Serialization/#using-other-validation-libraries
+ ../Reference/Validation-and-Serialization.md#using-other-validation-libraries

docs/Tutorial/07-hooks.md
- /docs/latest/Reference/Hooks/
+ ../Reference/Hooks.md
- /docs/latest/Reference/Lifecycle/#lifecycle
+ ../Reference/Lifecycle.md#lifecycle
```

This is not a competing bounty claim; just leaving the exact CI failure map so the existing PR can move forward with less review churn.

ai-hustle-bro commented:
I also opened a small helper PR against the source branch with those linkChecker fixes applied: https://github.com/jean-michelet/fastify/pull/2

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

jean-michelet commented:
I take back this PR, I can split the bounty with a dedicated reviewer from @fastify/collaborators

jean-michelet committed:
docs(tutorial): add test app helper

jean-michelet committed:
docs(tutorial): add configuration chapter

jean-michelet committed:
docs(tutorial): add database integration with Knex

jean-michelet committed:
Add CORS tutorial chapter

jean-michelet committed:
Add rate limiting tutorial chapter

jean-michelet committed:
Add authentication and authorization tutorial

jean-michelet committed:
docs(tutorial): organize application plugins by domain

jean-michelet committed:
docs(tutorial): add user registration chapter

jean-michelet committed:
docs(tutorial): separate authentication and authorization

jean-michelet committed:
docs(tutorial): reuse Redis for rate limiting

jean-michelet committed:
docs(tutorial): refresh the tutorial introduction

jean-michelet committed:
docs(tutorial): restore foundational Fastify explanations

jean-michelet committed:
docs(tutorial): organize application tests and plugins

jean-michelet committed:
docs(tutorial): align database and CORS integration

jean-michelet committed:
docs(tutorial): complete registration and session authentication

jean-michelet committed:
docs(tutorial): clarify authorization and rate limiting

jean-michelet commented:
The Swagger and TypeScript/Architecture parts are still missing and I need to review them again.
But there are already 18 chapters written.

Can you tell me what you tell me what you can review @fastify/core @fastify/plugins?

Honestly, adding some chapters about real-world deployments and constraints could give a book.

jean-michelet committed:
Merge branch 'main' into docs/tutorials-part-1

mcollina commented on docs/Tutorial/03-create-server.md:
```diff
@@ -0,0 +1,164 @@
+
+## Write the server code
+
+Create a new file named `server.js` in your project directory and
```
Given the community sentiment, we should use typescript with type stripping for this guide.

jean-michelet replied:
Agree, don't remember why I've written it in vanilla JS.

jean-michelet replied:
I know that AI agents are performing some refactoring pretty well these days, because this comment doesn’t make me feel clinically depressed.

jean-michelet replied:
Cooking :robot:

jean-michelet committed:
docs(tutorial): use native TypeScript type stripping

jean-michelet committed:
docs(tutorial): infer route types with TypeBox

jean-michelet commented:
Will do my best to achieve typing features we want before publishing this tutorial, but you can still review some ideas.

jean-michelet committed:
docs(tutorial): add OpenAPI documentation chapter

jean-michelet committed:
docs(tutorial): add architecture tools chapter

jean-michelet committed:
docs(tutorial): remove architecture tools chapter

metcoder95 commented on docs/Tutorial/03-create-server.md:
````diff
@@ -0,0 +1,193 @@
+
+```bash
+npm init -y
+npm install fastify@5
````
Are we gonna add this to the checklist so we refresh the tutorial upon new majors?

jean-michelet replied:
No preference, we can also remove the major constraint, I think the tutorial will still be valid in v6.

metcoder95 commented on docs/Tutorial/06-validation-and-serialization.md:
```diff
@@ -0,0 +1,491 @@
+We’ll move all schemas into a `schemas.ts` file, configure Ajv with safe
+defaults, and attach schemas to all our routes.
+
+### `schemas.ts`
```
I'm not opposed to using alternatives for Input Validation, but as this tutorial is made to teach `fastify` from scratch, shouldn't be better to just stick to what `fastify` ships and add sections showcasing how alternatives for each feature can be used?

> Just what you did lines above

jean-michelet replied:
It was my initial intention, but as @mcollina suggest we write it directly in typescript.
TypeBox is really the recommended default according to our guide: https://fastify.dev/docs/latest/Reference/TypeScript/#json-schema

metcoder95 replied:
No strong opinion, wanted to raise that as we were aiming for an initial guide 👍

metcoder95 commented on docs/Tutorial/08-errors.md:
````diff
@@ -0,0 +1,186 @@
+});
+```
+
+### The `@fastify/error` package
````
Don't forget: https://github.com/fastify/fastify-sensible

metcoder95 commented on docs/Tutorial/09-plugins-and-encapsulation.md:
```diff
@@ -0,0 +1,745 @@
+
+Plugins are cohesive, reusable functions that extend a Fastify
+application. A plugin receives a scoped Fastify instance. Inside that
+scope it can add routes, decorators, hooks, and any other Fastify
```
```suggestion
within that function it can add routes, decorators, hooks, and any other Fastify
```

Just trying to avoid collapse the `scope` concept as I've seen is quite confusing for people using the framework for the first time

jean-michelet replied:
Ok, I will try to make it more clear, but still teach the encapsulation scope/context at play.

metcoder95 commented on docs/Tutorial/10-testing.md:
````diff
@@ -0,0 +1,497 @@
+Install borp as a development dependency:
+
+```bash
+npm i -D borp
````
Not opposed to `borp`, but what about plain `node --test`; you can take a look at how is used in different projects

jean-michelet replied:
I think it's a nice tool for testing ts code with built-in coverage

jean-michelet commented:
@metcoder95 Thanks for the feedback!

Working at at auth0 you might have some expertise in authentication?
I expected some feedback from you regarding my Redis-Backed session implementation, authorization and
the arguments I develop.

mcollina commented:
## Review of PR #6239 – Fastify Tutorial (Guide Feedbacks)

*This review was conducted using **DeepSeek V4 Flash** via the pi-coding-agent harness.*

---

### Overview

I checked out the PR branch, followed the guide chapter by chapter, and tested all examples in `~/tmp/fastify-guide/`. Below are my findings, organized by chapter, with a strong emphasis on the need for a **full TypeScript version of this guide**.

---

### Critical: The Guide Must Be Available in TypeScript

This is the most important feedback. The entire tutorial is written in JavaScript (ESM), but:

1. **Fastify has first-class TypeScript support** — `fastify.d.ts`, `@types/fastify`, and the community standard is TS.
2. **Newcomers to Fastify who use TypeScript cannot directly follow this guide** — every code snippet needs mental translation to TS types, generics, and decorator patterns.
3. **The PR adds 18 chapters (~7,800 lines) of docs** — without a TypeScript companion, the largest documentation update in Fastify's history is inaccessible to a huge portion of its user base.

**Recommendation**: Either:
- Provide a `docs/Tutorial/typescript/` parallel tree with `.ts` examples, OR
- Add TypeScript callouts/side-by-side snippets in each chapter, OR
- At minimum, add a prominent note explaining that the JS examples can be typed with Fastify's generic type parameters (`<Body, Params, Query, Headers>`).

---

### Chapter-by-Chapter Findings

#### Chapters 03–10 (Core Concepts)

All examples tested successfully. A few notes:

- **Chapter 06** (`app.listen({ port })` vs `{ host, port }`) is inconsistent across chapters.
- **Chapter 08**: `setNotFoundHandler` must be called after all routes are registered — worth documenting.
- **Chapter 10**: 100% coverage with `borp` works well; good pattern.

#### Chapter 11 – Organizing Application Plugins (⚠️ Bugs Found)

Three significant issues:

1. **Authentication hooks not shared**: `authenticationHooksPlugin` is registered as a child of `authenticationPlugin` but is NOT wrapped with `fastify-plugin`. Since `authenticationPlugin` skips encapsulation (`fp`), its child creates an encapsulated sibling scope that doesn't share hooks with `quotesPlugin`. Fix: wrap `authenticationHooksPlugin` with `fp()`.

2. **`db` decorator inaccessible at root**: `quotesDatabaseServicePlugin` (which is `fp`) is registered inside `quotesPlugin` (which has `encapsulate: true`). The `db` decorator is promoted only to `quotesPlugin`'s scope, not to the application root. Tests that access `app.db` at root scope fail.

3. **Missing `addSchema(idParam)`**: In `quotes.routes.js`, `idParam` is used via `{ $ref: "idParam#" }` but never registered with `app.addSchema()`. Only `quoteResponse` and `errorMessage` are registered. Causes schema resolution failures.

These bugs would be more obvious in TypeScript because the type system would flag missing decorators and inaccessible properties at compile time.

#### Chapter 13 – Database Integration with Knex

Works correctly with Docker Compose PostgreSQL. However:

- **Recommendation**: Consider using **Drizzle** or **Prisma** instead of Knex. Most modern Node.js/TypeScript projects use these tools. Knex is battle-tested but feels dated.
- Tests share `quote_vault` database with the dev server — use `quote_vault_test` instead.

#### Chapter 14 – CORS

CORS test expects `undefined` header for non-allowed origins, but `@fastify/cors` returns the configured origin regardless. The browser enforces the mismatch, not the server. The test needs adjustment.

#### Chapters 15–18 (Auth, Authorization, Rate Limiting)

Reviewed but not fully tested (require Redis). Notable issues:

- Password policy (min 12 chars, upper+lower+digit+symbol) is very strict for a demo.
- Chapter 16 replaces the fake auth plugin with real sessions — the transition is complex and should explicitly state that the old plugin must be completely removed.
- Chapter 18 likely has a **copy-paste error**: the rate-limit key function references `app.knex` instead of the Redis client.

---

### TypeScript-Specific Recommendations

If the guide is rewritten in TypeScript, these patterns should be documented:

```typescript
import Fastify, { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify'
import { Type, Static } from '@sinclair/typebox'  // or zod

// Typed schemas with TypeBox
const QuoteBody = Type.Object({
  text: Type.String({ minLength: 1 })
})
type QuoteBodyType = Static<typeof QuoteBody>

// Typed routes
app.get<{ Params: { id: number }; Reply: QuoteResponse }>('/quotes/:id', ...)

// Typed decorators
app.decorate('db', createDb())
declare module 'fastify' {
  interface FastifyInstance {
    db: ReturnType<typeof createDb>
  }
}

// Typed hooks with request/reply generics
app.addHook('onRequest', async (request: FastifyRequest, reply: FastifyReply) => { ... })
```

---

### Minor Issues

- Port 3000 is used throughout — easy to get EADDRINUSE between chapters.
- The guide should mention `pkill -f "node server"` or port rotation.
- `close-with-grace` is a good pattern but `app.close()` with `forceCloseConnections: false` is already built into Fastify 5.

---

### Summary

This is a **massive, well-structured tutorial** that fills a real gap in Fastify's documentation. However, it has:

- **3 encapsulation bugs** in chapter 11 that break the plugin architecture
- **No TypeScript version**, making it inaccessible to TS-first developers
- **A dated database recommendation** (Knex vs Drizzle/Prisma)
- **Minor test and consistency issues** across chapters

Fixing the encapsulation bugs and providing a TypeScript companion would make this PR production-ready.

---

*Full detailed feedback with per-chapter test results is available in `./guide_feedbacks.md` in the repository.*

jean-michelet commented:
I think you ran your review based on an earlier version. I’m using TS now.

Some of these bugs are real, or more precisely, drafting errors that are corrected as the chapters have progressed.

Like you I ran an agent that impersonate a reader building the app from the chapters.

I am fixing that.

jean-michelet committed:
docs(tutorial): fix plugin refactor progression

Tony133 commented:
I think it would be great if we could have the code snippets written in both JS and TS. Other than that, it looks like a great tutorial! Having both would make it more comprehensive.

jean-michelet committed:
docs(tutorial): clarify plugin encapsulation terminology

jean-michelet commented:
> I think it would be great if we could have the code snippets written in both JS and TS

We'll see that later, maybe with the new website we could add this kind of feature by parsing the markdown.

jean-michelet committed:
Merge branch 'fastify:main' into docs/tutorials-part-1

jean-michelet committed:
docs(tutorial): fix offline reference links

metcoder95 commented on docs/Tutorial/16-authentication.md:
```diff
@@ -0,0 +1,1240 @@
+## How sessions work
+
+After login, the server sends a random session identifier in a signed cookie.
+The browser returns that cookie on later requests. User data remains in the
```
Maybe we can also the difference between user flow (like a browser) and a machine-to-machine (Bearer token).

Not need to develop that much but just make a mention.

metcoder95 commented on docs/Tutorial/16-authentication.md:
```diff
@@ -0,0 +1,1240 @@
+  async function authenticationHookPlugin (app) {
+    app.addHook(
+      'onRequest',
+      async function authenticationHook (request, reply) {
```
What do you think about abstracting this into the plugin? Just to make the plugin decorate the instance with `fastify.authenticate` so it can be use only when required.

Another option is to create such hook, and read the `routeOptions.config` to hint when to authenticate and when not to?

jean-michelet replied:
> What do you think about abstracting this into the plugin? Just to make the plugin decorate the instance with fastify.authenticate so it can be use only when required.

Ok.

metcoder95 reviewed (commented):
overall the authn/z lgtm, just small comments. Nice work!
